File: 0382-Linked List Random Node/0382-Linked List Random Node.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
 class Solution {
public:
    /** @param head The linked list's head.
        Note that the head is guaranteed to be not null, so it contains at least one node. */
    Solution(ListNode* head) {
        this->head = head;
    }
    
    /** Returns a random node's value. */
    int getRandom() {
        int result = head->val;
        ListNode* node = head->next;
        int i = 2;
        while (node)
        {
            if (rand() % i == 0)
            {
                result = node->val;
            }
            
            ++i;
            node = node->next;
        }
        
        return result;
    }
    
private:
    ListNode* head;
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(head);
 * int param_1 = obj.getRandom();
 */
 

File: 0301-0400/0382-Linked List Random Node/0382-Linked List Random Node.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None
import random
class Solution:

    def __init__(self, head: ListNode):
        """
        @param head The linked list's head.
        Note that the head is guaranteed to be not null, so it contains at least one node.
        """
        self.head = head

    def getRandom(self) -> int:
        """
        Returns a random node's value.
        """
        currLen = 1
        result = self.head.val
        curr = self.head.next
        while curr:
            currLen += 1
            if random.randrange(currLen) == 0:
                result = curr.val
            curr = curr.next
            
        return result

# Your Solution object will be instantiated and called as such:
# obj = Solution(head)
# param_1 = obj.getRandom()
